bundles/framework/layerlist/view/LayerViewTabs/LayerList/LayerCollapse/util.js

```diff
@@ -45,26 +45,66 @@ export const groupLayers = (layers, method, tools, allGroups = [], allDataProvid
         return typeof groupId === 'number' ? groupId : -Math.random();
     };
 
+    const mapGroupsAndLayers = function (group) {
+        // group has subgroups
+        if (group.hasSubgroups()) {
+            // check that group has layers
+            if (group.layers && group.layers.length > 0) {
+                const lang = Oskari.getLang();
+                const name = group.getName();
+                let newGroup = Oskari.clazz.create(
+                    'Oskari.mapframework.bundle.layerselector2.model.LayerGroup',
+                    group.id, method, name[lang]
+                );
+                newGroup.setTools(tools);
+                // add layer(s) to group
+                for (var i in group.layers) {
+                    layers.sort((a, b) => comparator(a, b, method))
+                        .filter(layer => !layer.getMetaType || layer.getMetaType() !== 'published')
+                        .forEach(layer => {
+                            if (group.layers[i].id == layer._id) {
+                                newGroup.addLayer(layer);
+                            }
+                        });
+                }
+                // check possible subgroups
+                const mappedSubgroups = group.getGroups().map(subgroup => {
+                    return mapGroupsAndLayers(subgroup);
+                });
+                // filter out subgroups that don't have a layer nor a subgroup with a layer
+                const filteredSubGroups = mappedSubgroups.filter(g => g !== undefined);
+                filteredSubGroups.length > 0 && newGroup.setGroups(filteredSubGroups);
+                return newGroup;
+            }
+        } else {
+            if (group.layers && group.layers.length > 0) {
+                const subLang = Oskari.getLang();
+                const subName = group.getName();
+                let newGroup = Oskari.clazz.create(
+                    'Oskari.mapframework.bundle.layerselector2.model.LayerGroup',
+                    group.id, method, subName[subLang]
+                );
+                newGroup.setTools(tools);
+                for (var j in group.layers) {
+                    layers.sort((a, b) => comparator(a, b, method))
+                        .filter(layer => !layer.getMetaType || layer.getMetaType() !== 'published')
+                        .forEach(layer => {
+                            if (group.layers[j].id == layer._id) {
+                                newGroup.addLayer(layer);
+                            }
+                        });
+                }
+                return newGroup;
+            }
+        }
+    };
```

Comment:
Actually most of the method is duplicated for some reason. The only thing that seems to be different in the ifs is recursion that's moved to the end of the method in this impl:
```suggestion
    const mapGroupsAndLayers = function (group) {
        if (!group.layers || group.layers.length === 0) {
            return;
        }
        const lang = Oskari.getLang();
        const name = group.getName();
        let newGroup = Oskari.clazz.create(
            'Oskari.mapframework.bundle.layerselector2.model.LayerGroup',
            group.id, method, name[lang]
        );
        newGroup.setTools(tools);
        // add layer(s) to group
        for (var i in group.layers) {
            layers.sort((a, b) => comparator(a, b, method))
                .filter(layer => !layer.getMetaType || layer.getMetaType() !== 'published')
                .forEach(layer => {
                    if (group.layers[i].id == layer._id) {
                        newGroup.addLayer(layer);
                    }
                });
        }
        if (!group.hasSubgroups()) {
            return newGroup;
        }
        // group has subgroups
        // check possible subgroups
        const mappedSubgroups = group.getGroups().map(subgroup => {
            return mapGroupsAndLayers(subgroup);
        });
        // filter out subgroups that don't have a layer nor a subgroup with a layer
        newGroup.setGroups(mappedSubgroups.filter(g => g !== undefined));
        return newGroup;
    };
```

Reply:
What is the layer object here though where you refer to its id with an underscore "private" variable `layer._id`? Is it an instance of an AbstractLayer? You should use getter instead.